MP Rest Client - meaningless exceptions

### Describe the bug

I'm using MicroProfile Rest Client with Quarkus and I have an issue with rest client exceptions.

My use case is simple.

If I make mistake in rest client URL I'll get 404 and following stacktrace

```
2021-12-09 21:33:16,635 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: org.jboss.resteasy.reactive.ClientWebApplicationException: Unknown error, status code 404
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:145)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:320)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:164)
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:251)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: javax.ws.rs.WebApplicationException: Unknown error, status code 404
        at io.quarkus.rest.client.reactive.runtime.DefaultMicroprofileRestClientExceptionMapper.toThrowable(DefaultMicroprofileRestClientExceptionMapper.java:17)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientResponseFilter.filter(MicroProfileRestClientResponseFilter.java:36)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:20)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:9)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
        ... 11 more
```

This stacktrace is totally meaningless.

After reading this stacktrace I don't know:

* Which of rest clients failed  ```MyRestClient```
* What rest client method was called  ```MyRestClient.apiCall() or https://my-api:8080/endpoint```
* Where it was called from ```MyResource.callAPI() -> MySerivice.apiCall() -> MyRestClient.apiCall()```

With ExceptionMapper I can extract rest client endpoint details from ClientWebApplicationException. Unfortunately I can't find any details about the call side.

How can I extract full stacktrace of the call side like "MyResource.callAPI() -> MySerivice.apiCall() -> MyRestClient.apiCall()".
I've tried creating my own ResponseExceptionMapper but I can't find nothing about the call side there.

At the moment only try-catch at the call side can give me full details about rest client call side.


### Expected behavior

Quarkus should provide tdetails about rest client exception in stacktace out of the box - it's rather basic use case

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java
@@ -10,6 +10,7 @@ import io.vertx.core.http.HttpClientResponse;
 import java.io.IOException;
 import java.io.InputStream;
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
 import java.net.URI;
@@ -46,6 +47,8 @@ import org.jboss.resteasy.reactive.spi.ThreadSetupAction;
  */
 public class RestClientRequestContext extends AbstractResteasyReactiveContext<RestClientRequestContext, ClientRestHandler> {
 
+    private static final String MP_INVOKED_METHOD_PROP = "org.eclipse.microprofile.rest.client.invokedMethod";
+
     private final HttpClient httpClient;
     // Changeable by the request filter
     String httpMethod;
@@ -64,6 +67,8 @@ public class RestClientRequestContext extends AbstractResteasyReactiveContext<Re
     private final boolean checkSuccessfulFamily;
     private final CompletableFuture<ResponseImpl> result;
     private final ClientRestHandler[] abortHandlerChainWithoutResponseFilters;
+
+    private final boolean disableContextualErrorMessages;
     /**
      * Only initialised if we have request or response filters
      */
@@ -129,6 +134,10 @@ public class RestClientRequestContext extends AbstractResteasyReactiveContext<Re
         this.result = new CompletableFuture<>();
         // each invocation gets a new set of properties based on the JAX-RS invoker
         this.properties = new HashMap<>(properties);
+
+        // this isn't a real configuration option because it's only really used to pass the TCKs
+        disableContextualErrorMessages = Boolean
+                .parseBoolean(System.getProperty("quarkus.rest-client.disable-contextual-error-messages", "false"));
     }
 
     public void abort() {
@@ -140,8 +149,16 @@ public class RestClientRequestContext extends AbstractResteasyReactiveContext<Re
     protected Throwable unwrapException(Throwable t) {
         var res = super.unwrapException(t);
         if (res instanceof WebApplicationException) {
-            WebApplicationException webApplicationException = (WebApplicationException) res;
-            return new ClientWebApplicationException(webApplicationException.getMessage(), webApplicationException,
+            var webApplicationException = (WebApplicationException) res;
+            var message = webApplicationException.getMessage();
+            var invokedMethodObject = properties.get(MP_INVOKED_METHOD_PROP);
+            if ((invokedMethodObject instanceof Method) && !disableContextualErrorMessages) {
+                var invokedMethod = (Method) invokedMethodObject;
+                message = "Received: '" + message + "' when invoking: Rest Client method: '"
+                        + invokedMethod.getDeclaringClass().getName() + "#"
+                        + invokedMethod.getName() + "'";
+            }
+            return new ClientWebApplicationException(message, webApplicationException,
                     webApplicationException.getResponse());
         }
         return res;

--- a/tcks/microprofile-rest-client-reactive/pom.xml
+++ b/tcks/microprofile-rest-client-reactive/pom.xml
@@ -30,6 +30,7 @@
                         <quarkus.arc.remove-unused-beans>false</quarkus.arc.remove-unused-beans>
                         <quarkus.rest-client-reactive.scope>javax.enterprise.context.Dependent</quarkus.rest-client-reactive.scope>
                         <quarkus.rest-client-reactive.disable-smart-produces>true</quarkus.rest-client-reactive.disable-smart-produces>
+                        <quarkus.rest-client.disable-contextual-error-messages>true</quarkus.rest-client.disable-contextual-error-messages>
                         <vertx.disableTCCL>true</vertx.disableTCCL>
                         <wiremock.server.port>${wiremock.server.port}</wiremock.server.port>
                     </systemPropertyVariables>

